Library/Homebrew/brew.sh

```diff
@@ -195,7 +195,8 @@ then
   HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh"
 fi
 
-if [[ "$(id -u)" = "0" && "$(/usr/bin/stat -f%u "$HOMEBREW_BREW_FILE")" != "0" ]]
+brew_file_ls_info=($(ls -nd "$HOMEBREW_BREW_FILE"))
+if [[ "$(id -u)" = "0" && "${brew_file_ls_info[2]}" != "0" ]]
```

Comment:
Can we move this check into a shell function? This would allow for early returns (avoiding an `ls` invocations who's output will never be used in the typical case) and local variables (avoiding additional global variables like `brew_file_ls_info`).

For performance reasons it would also be desirable to exchange the inner and outer check, i.e. first check if we've been invoked with a command that we prefer not to run as root, and only if that's the case, check whether we're root and `bin/brew` is owned by root.

Here's a suggestion (untested), that also incorporates my desire to avoid parsing `ls` output:

```sh
check-run-command-as-root() {
  case "$HOMEBREW_COMMAND" in
    analytics|\
    create|\
    install|\
    link|\
    migrate|\
    pin|\
    postinstall|\
    reinstall|\
    switch|\
    tap-pin|\
    tap|\
    update|\
    upgrade|\
    vendor-install)
      # Continue check as we prefer to not run these as root.
      ;;
    *)
      return
      ;;
  esac

  [[ "$(id -u)" = 0 ]] || return

  # Assume BSD user space, but fall back to GNU if this fails.
  local brew_file_uid
  brew_file_uid="$(/usr/bin/stat -f%u "$HOMEBREW_BREW_FILE" 2>/dev/null)"
  : "${brew_file_uid:="$(stat --format=%u "$HOMEBREW_BREW_FILE" 2>/dev/null)"}"
  [[ "$brew_file_uid" != 0 ]] || return

  odie <<EOS
Cowardly refusing to 'sudo brew $HOMEBREW_COMMAND'
You can use brew with sudo, but only if the brew executable is owned by root.
However, this is both not recommended and completely unsupported so do so at
your own risk.
EOS
}

check-run-command-as-root
```